### ISSUE TITLE
List groups returns empty.. No other way to join a group

### ISSUE THREAD

**Issue by lazynomad** [bbernhard/signal-cli-rest-api#127]
ListGroups have been returning empty even though the account belongs to a group.

Lets say I use "userX" to run signal-Cli. It's registered and verified with your docker. Sending direct messages working great.
My secondary number is "userY", which I use on the Signal app on mobile.

- I created a new group on mobile as "userY".
- Added "userX" to the group.
- Got a message that "userX" must accept the invite

After this step, nothing could be done. "ListGroups" as "userX" returns empty list (is this because the "userX" hasn't accepted the invite yet) ?
"userX" cannot join this group without knowing it's groupId. 
"userY" can generate invite link but I don't see any support for accepting invite link in your api.

Please help.

**Comment by bbernhard** [comment]
Please execute the `receive` endpoint (`curl -X GET -H "Content-Type: application/json" 'http://127.0.0.1:8080/v1/receive/<your signal number>'`) and then try the list groups endpoint again.

**Comment by lazynomad** [comment]
You are awesome.. That worked. 
Thank you. :)

**Comment by bbernhard** [comment]
You're welcome!

btw: [signal-cll](https://github.com/AsamK/signal-cli) which this wrapper is based on) recommends in its man page to call `receive` on a regular basis. I would recommend to create a cronjob (or something similar) and call the `receive` endpoint once a day/every few days. :)

**Comment by lazynomad** [comment]
Sure will do that.. 
1) By the way, do you know if there's some kind of throttling on the Signal endpoint on the number of messages sent per second ?

2) Also, do you use DBus underneath or call the Signal-Cli directly ?

**Comment by bbernhard** [comment]
Regarding 1: The slowness you are probably experiencing when sending messages is (partly) due to to slow JVM startup time. Especially on ARM based devices (like the Raspberry Pi) sending messages is pretty slow. 

Regarding 2: I am using the signal-cli binary directly. Mainly because of two reasons: 

*) not all commands are exposed via the DBUS interface
*) some commands (e.g the `receive`) do not map well to "push/pull based" interfaces (like the REST API interface). e.g: If there are some messages received on the DBUS, we would need to cache/store them somewhere, so that the messages can be returned when the user calls the `receive` endpoint again. It's definitely not an unsolvable problem, but it needs a quite a bit of thought to get right (one of the problems that needs to be solved is how those messages are stored then. As Signal uses E2E Encryption it feels kind of bad to receive those messages and store them in plaintext somewhere.)

signal-cli recently added the experimental "native mode" which compiles the application to a native image (this should reduce the JVM startup time). in case you want to read up on the discussion regarding the native mode, have a look here: https://github.com/AsamK/signal-cli/issues/446. The native mode is still experimental, but it already works pretty well on x86-64 systems. (on a related note: A few days ago a pull request was merged into signal-cli's master branch which fixed some bugs with the native mode; see here: https://github.com/AsamK/signal-cli/pull/576).

**Comment by lazynomad** [comment]
Great points. Wasn't aware of this design flaw when we use Dbus interface. 

On point (1), my question is with respect to the Signal endpoint throttling but not the Signal-Cli. I could spin up multiple threads to send messages in parallel. However if Signal endpoint throttles our requests, then I would have to add back-off. Since there I couldn't find any official documentation on the Signal apis, wondering if you ever experienced rate limiting ?

**Comment by bbernhard** [comment]
> Great points. Wasn't aware of this design flaw when we use Dbus interface.
> 
> On point (1), my question is with respect to the Signal endpoint throttling but not the Signal-Cli. I could spin up multiple threads to send messages in parallel. However if Signal endpoint throttles our requests, then I would have to add back-off. Since there I couldn't find any official documentation on the Signal apis, wondering if you ever experienced rate limiting ?

I think that wouldn't work - at least not in parallel. signal-cli stores all its config (contacts, keys, etc) in a config file. When the signal-cli binary is invoked it locks the config file  to prevent multiple processes accessing the file and corrupting it. I am not sure anymore which type of file lock it uses. If it's a blocking lock it might be possible to invoke multiple signal-cli instances in parallel and let them wait for the lock. But I haven't tried that.

Regarding the Signal Server Rate limiting: I am pretty sure that there is a rate limit (at least I would configure one if I would set up such a server). But for me the JVM startup time was always the bottleneck...so I never really hammered the server in the first place.

**Comment by lazynomad** [comment]
Just want to say Thank you for your help.. Managed to write an application that's used to scan for Covid-19 vaccine availability in India. Currently the situation is dire in India. 
I added reference to your repo as this couldn't have been possible without your awesome work. You simplified the usage of Signal-Cli as I had too many problems getting the Signal-Cli natively running with Dbus. 

[Link to repo](https://github.com/lazynomad/cowin-signal)

**Comment by lazynomad** [comment]
By the way, do you know if I can run two Signal-Cli instances associated with same mobile number at the same time ? I remember reading Signal treats multiple nodes as Main and child nodes, where all of them could be active at the same time using the same mobile number. Did you ever try that ? I am wondering if I could leverage that for load balancing.

**Comment by bbernhard** [comment]
Many thanks - it's always awesome to hear when people put open source software to good use! :) 

> By the way, do you know if I can run two Signal-Cli instances associated with same mobile number at the same time ?

I think that probably won't work for the reasons I've described above. `signal-cli` uses a config file to store data associated with that number (e.g keys, some profile data, etc). Any REST API call that has the mobile number in the URL parameter invokes `signal-cli` with the `-u <phone number>` parameter. `signal-cli` then uses a file lock to prevent multiple `signal-cli` instances from accessing the same config file (I guess that's to avoid any file corruption when multiple instances want to update the config file at the same time). 

What you could try:

You could copy the `signal-cli-config` folder (if you are using the example `docker-compose.yml` file from the README.md, then the `signal-cli-config` folder will be in the same directory as your `docker-compose.yml` file resides) to a different location and start another docker container which uses the copy. I haven't actually tried & verified it, but my hope is that not every `signal-cli` actually modifies the config file. So, if you are just using the `send` endpoint, it might work to do that on two different copies. 

If you perform a REST call that modifies the config file (e.g: add someone to a group or do a `receive`) you should that only with your "primary" docker container (maybe stop all the other "secondary" containers in the meantime). Once the operation is performed, you could distribute the modified `signal-cli` config files to all the other containers and start them again. So you could basically write your own "orchestration layer" (a simple python/bash script) on top of all your worker nodes. 

I've no idea whether that actually works (just had that idea a few seconds ago), so maybe it's complete bullshit. ;) Of course that only works in case `signal-cli` doesn't change the config when doing a `send`. 

If that actually works I think this might be a cool idea to scale horizontally :thinking:. (at least when one wants to optimize for sending). But it sounds a bit too good to be true :D 

An alternative would be to have a look at the native mode. (see the `README.md` file on how to enable that one) At least on x86-64 systems that one should work quite well.

**Comment by bbernhard** [comment]
Oh damn, looks like `signal-cli` changes the config file even when sending something. There's a `record` parameter in there which will be updated when the message was sent :( 

Just started two docker instances with separate config files listening on two different ports and invoked the `send` endpoint for both containers at the same time. Both messages were successfully sent to my mobile device, but unfortunately I only received one :/

**Comment by lazynomad** [comment]
Thanks for trying it out and reporting. That's a bummer. I also see that they recently exposed a socket interface.. would be really cool to explore, when you got time.

By the way, if I had to spin up another container and Link it to the first container signal-cli instance, do you have any api exposed for that ? I see your swagger config shows "/v1/qrcodelink". This makes sense if I am scanning a qr code from my mobile but if the "master" is another signal-cli instance, I don't see any API to POST the generated tsdevice:/ string..

**Comment by bbernhard** [comment]
> I also see that they recently exposed a socket interface.. would be really cool to explore, when you got time.

Do you have a link to that one? That would be interesting. I only know that `signald` has a socket interface, but wasn't aware that `signal-cli` has one too.

I guess what you are looking for is a REST API wrapper for the `link` and `addDevice` signal-cli commands, right? Unfortunately, I do not use any linked devices, so I can't really test the functionality. But it should be quite easy to extend the `/v1/qrcodelink` endpoint to alternatively return the `tsdevice` string instead of the QR code and to add another endpoint `signal-cli`'s  `addDevice` command. So in case you need that functionality, just let me know. I can quickly add that.

btw: In case you want to run `signal-cli` commands directly, you can do it like that:

* run `docker ps` to get the container id
* connect to the running container with `docker exec -it <containerid> /bin/bash`
* switch to the signal user with `su signal-api`
* you can now run every `signal-cli` command you want. Just make sure to always specify the config file with `--config /home/.local/share/signal-cli`.

This allows you to quickly test some `signal-cli` commands where there's no REST API endpoint yet. 

fyi: When you've already registered your number with `signal-cli` you might need to remove your  `signal-cli` config first and start from scratch, otherwise you might experience some strange issues. see also this post [here](https://github.com/bbernhard/signal-cli-rest-api/issues/35#issuecomment-759380739)
